File: test/languages/sample.rb
# frozen_string_literal: true

module Akihabara
  Review = Data.define(:id, :title, :score, :state) do
    def published? = state == :published

    def summary
      formatted = score ? format('%.1f', score) : 'unrated'
      "#{title}: #{formatted}"
    end
  end

  class Catalog
    DEFAULT_LIMIT = 10

    def initialize(reviews)
      @reviews = reviews.freeze
    end

    def summaries(limit: DEFAULT_LIMIT)
      @reviews
        .select(&:published?)
        .sort_by { |review| -(review.score || 0) }
        .first(limit)
        .map(&:summary)
    rescue ArgumentError => error
      warn error.message
      []
    end
  end
end

puts Akihabara::Catalog.new([Akihabara::Review.new(1, 'Night Walk', 8.5, :published)]).summaries

File: test/languages/sample.lua
local DEFAULT_LIMIT <const> = 10

---@class Review
---@field id integer
---@field title string
---@field score number?
local Review = {}
Review.__index = Review

function Review.new(id, title, score)
    return setmetatable({ id = id, title = title, score = score }, Review)
end

function Review:summary()
    local score = self.score and string.format("%.1f", self.score) or "unrated"
    return string.format("%d: %s (%s)", self.id, self.title, score)
end

local reviews = { Review.new(1, "Night Walk", 8.5) }
for index, review in ipairs(reviews) do
    if index <= DEFAULT_LIMIT then
        print(review:summary())
    end
end
